src/test/java/htsjdk/variant/bcf2/BCF2WriterUnitTest.java

```diff
@@ -0,0 +1,283 @@
+        ret.addMetaDataLine(new VCFInfoHeaderLine("DP", 1, VCFHeaderLineType.String, "x"));
+        ret.addMetaDataLine(new VCFInfoHeaderLine("BB", 1, VCFHeaderLineType.String, "x"));
+        ret.addMetaDataLine(new VCFInfoHeaderLine("GQ", 1, VCFHeaderLineType.String, "x"));
+        ret.addMetaDataLine(new VCFFormatHeaderLine("GT", 1, VCFHeaderLineType.String, "x"));
+        ret.addMetaDataLine(new VCFFormatHeaderLine("DP", 1, VCFHeaderLineType.String, "x"));
+        ret.addMetaDataLine(new VCFFormatHeaderLine("BB", 1, VCFHeaderLineType.String, "x"));
+        ret.addMetaDataLine(new VCFFormatHeaderLine("GQ", 1, VCFHeaderLineType.String, "x"));
+        ret.setSequenceDictionary(sequenceDict);
+        return ret;
+    }
+
+    @BeforeClass
+    private void createTemporaryDirectory() {
+        File tempDir = TestUtil.getTempDirectory("BCFWriter", "StaleIndex");
+        tempDir.deleteOnExit();
+    }
+
+
+    /**
+     * test, using the writer and reader, that we can output and input BCF without problems
+     */
+    @Test
+    public void testWriteAndReadBCF() throws IOException {
+        final File bcfOutputFile = VariantBaseTest.createTempFile("testWriteAndReadVCF.", ".bcf");
+
+        Tribble.indexFile(bcfOutputFile).deleteOnExit();
+        Set<VCFHeaderLine> metaData = new HashSet<VCFHeaderLine>();
+        Set<String> additionalColumns = new HashSet<String>();
+        final SAMSequenceDictionary sequenceDict = createArtificialSequenceDictionary();
+        final VCFHeader header = createFakeHeader(metaData, additionalColumns, sequenceDict);
+        try (final VariantContextWriter writer = new VariantContextWriterBuilder()
+                .setOutputFile(bcfOutputFile).setReferenceDictionary(sequenceDict)
+                .setOptions(EnumSet.of(Options.INDEX_ON_THE_FLY))
+                .build()) {
+            writer.writeHeader(header);
+            writer.add(createVC(header));
+            writer.add(createVC(header));
+        }
+        VariantContextTestProvider.VariantContextContainer container = VariantContextTestProvider
+                .readAllVCs(bcfOutputFile, new BCF2Codec());
+        int counter = 0;
+        final Iterator<VariantContext> it = container.getVCs().iterator();
+        while (it.hasNext()) {
+            it.next();
+            counter++;
+        }
+        Assert.assertEquals(counter, 2);
+
+
+        // prevent writing header twice
+        final VariantContextWriter writer2 = new VariantContextWriterBuilder()
+                .setOutputFile(bcfOutputFile).setReferenceDictionary(sequenceDict)
+                .setOptions(EnumSet.of(Options.INDEX_ON_THE_FLY))
+                .build();
+        writer2.writeHeader(header);
+        try {
+            writer2.writeHeader(header);
+            Assert.fail("Should not allow writing header twice");
+        } catch (Exception e) {
+            Assert.assertTrue(e instanceof IllegalStateException);
```

Comment:
This test (prevent writing the header twice) should be factored out into a separate test method that declares IllegalStateException as an expectedException in the test annotation (see example [here](https://github.com/samtools/htsjdk/blob/656dc24930e1dc7fd0abd4c5e2abedf6f5074b35/src/test/java/htsjdk/samtools/BAMFileWriterTest.java#L146)). It eliminates the need to write a try/catch block and cleanly separates the individual tests from each other.